classes/class-truelayer-encryption.php

```diff
@@ -87,104 +101,156 @@ public function __wakeup() {
 	}
 
 	/**
-	 * Update specific gateway settings.
+	 * Encrypt any values that should be encrypted before we store them in the database.
 	 *
 	 * @param array $settings The gateway settings.
 	 *
 	 * @return mixed
 	 */
-	public function update_truelayer_credit( $settings ) {
-		// if the key does not exist, don't do anything.
+	public function encrypt_values( $settings ) {
+		// If the key does not exist, don't do anything.
 		if ( ! defined( 'TRUELAYER_KEY' ) ) {
 			return false;
 		}
-		$private_key = null;
-		if ( 'yes' === $settings['testmode'] ) {
-			$private_key        = 'truelayer_sandbox_client_private_key';
-			$client_secret      = 'truelayer_sandbox_client_secret';
-			$client_certificate = 'truelayer_sandbox_client_certificate';
-		} else {
-			$private_key        = 'truelayer_client_private_key';
-			$client_secret      = 'truelayer_client_secret';
-			$client_certificate = 'truelayer_client_certificate';
 
-		}
-		$truelayer_options = array( $private_key, $client_secret, $client_certificate );
-		foreach ( $truelayer_options as $truelayer_option ) {
-			if ( ! empty( $this->settings[ $truelayer_option ] ) ) {
-				$old_value = $this->settings[ $truelayer_option ];
-				try {
-					if ( $this->decrypt( $settings[ $truelayer_option ] ) === $this->decrypt( $old_value ) ) {
-						$settings[ $truelayer_option ] = $old_value;
-					} else {
-						$settings[ $truelayer_option ] = $this->encrypt( $settings[ $truelayer_option ] );
-					}
-				} catch ( EnvironmentIsBrokenException | WrongKeyOrModifiedCiphertextException $e ) {
-					$settings[ $truelayer_option ] = $this->encrypt( $settings[ $truelayer_option ] );
-				}
-			} else {
-				$settings[ $truelayer_option ] = $this->encrypt( $settings[ $truelayer_option ] );
+		// Clear any encryption error notices. New errors will be generated if an issue exists.
+		delete_option( 'truelayer_encryption_error' );
+
+		// Loop each of the encrypted keys and ensure the value is encrypted before saving.
+		foreach ( $this->encrypted_keys as $encrypted_key ) {
+			// Get the current value of the setting before updating.
+			$current_value = $this->settings[ $encrypted_key ] ?? false;
+			$new_value     = $settings[ $encrypted_key ] ?? false;
+
+			// If the new value is empty, continue.
+			if ( empty( $new_value ) ) {
+				continue;
 			}
+
+			// Compare if the new value is the same as the current value after it is decrypted.
+			$are_equal = $this->are_equal( $current_value, $new_value );
```

Comment:
Question: when saving for the first time, the values won't be encrypted. `are_equal` will call `decrypt` which will fail and log an error message in `truelayer_encryption_error`. Won't this then show to the user in the UI?

Reply:
Potentially yes, good catch. Ill test it and make sure. I can't remember it being an issue during my tests, but it could be something i just missed. And looking at the code now again, it does look like a possibility yes.

Reply:
Thanks!

Reply:
I was unable to recreate the issue, however i added a check to not even attempt to decrypt an empty value from the current value instead. Changing the line `$decrypted_current_value = $this->decrypt( $current_value );` to `$decrypted_current_value = empty( $current_value ) ? '' : $this->decrypt( $current_value );` just to be safe.